docs/proposals/kubectl-apply-purge-from-namespace.md

```diff
@@ -0,0 +1,102 @@
+
+When user deletes some of the files or resources within files from `/etc/kubernetes/addons`, `kubectl apply --purge-from-namespace` would ensure the resources that were referenced in the deleted files are deleted, based on the assumption that `/etc/kubernetes/addons` represents all resources that are supposed to be present in `kube-system` namespace.
+
+
+## Managing Any Other Application
+
+It should be possible to apply the same approach to any other applications, with the assumption that single application occupies a namespace of its own.
+
+The usage may look like this:
+
+```
+kubectl apply --purge-from-namespace -f https://example.com/myapp1.json
+```
+
+Given resources in `myapp1.json` set their namespaces explicitly, resource that are no longer in this file would be purged, if no longer defined.
+
+```
+kubectl apply --purge-from-namespace -f https://example.com/myapp-part1.json -f https://example.com/myapp-part2.json
+```
+
+If `myapp1.json` and `myapp2.json` are in the same namespaces and it is explicitly defined, removing either of the URLs from the invocations arguments will result in all the resources that URL has defined to be purged.
+
+## Flag Discovery User Experience
+
+As stated above, the `--purge-from-namespace` flag shouldn't be enabled by default, however it should be discoverable.
+
+So, if a `kubectl apply` command is run without `--purge-from-namespace`, and the referenced files are all in the same namespace, then the command could output something like:
+
+```
+$ kubectl apply -f resources.yaml
+Updating ServiceA...
+Updating ServiceB...
+Updating DeploymentA...
+The following resources in namespace `foo` are not referenced in the files you provided:
+
+    DeploymentB
+    ServiceC
+
+You can automatically delete them, if you wish, using:
+
+    kubectl apply --purge-from-namespace -f resources.yaml
+
+$ kubectl apply --purge-from-namespace -f resources.yaml
+Updating ServiceA...
+Updating ServiceB...
+Updating DeploymentA...
+Deleting DeploymentB...
+Deleting ServiceC...
+Done!
+```
+
+If there are > 5 resources which match, we could just list the first 5.
+
+If no namespace, or conflicting namespaces are provided in the resource file(s), this output would not be shown.
+
+# Alternatives Considered
+
+A new Kubernetes API-level concept "Versioned ApplySets" could record all the `kubectl apply` commands that were ever executed against a cluster.
+This would allow deletions to be automatically detected and applied between invocations of `kubectl apply` without the onus of having to put resources into namespaces.
+
+This adds complexity, since a new Kubernetes API object type would need to be invented.
+
+It's not obvious how to make this concept work correctly in the case where one user runs `kubectl apply X` and later `kubectl apply X'` and a different user runs `kubectl apply Y` where `X'` is intended as an update to `X`
```

Comment:
Why is that a problem? The same issue exists already with a simple `kubectl apply` command if object names are re-used. If different users have different baselines for their changes, this asks for conflicts.